harfbuzz/src/buffer.rs

```diff
@@ -0,0 +1,308 @@
+/// assert_eq!(b.is_empty(), false);
+/// ```
+///
+/// ## Segment Properties
+///
+/// In addition to the text itself, there are three important properties
+/// that influence how a piece of text is shaped:
+///
+/// * Direction: The direction in which the output glyphs flow. This is
+///   typically left to right or right to left. This is controlled via
+///   the [`set_direction`] method on `Buffer`.
+/// * Script: Script is crucial for choosing the proper shaping behaviour
+///   for scripts that require it (e.g. Arabic) and the which OpenType
+///   features defined in the font to be applied. This is controlled via
+///   the [`set_script`] method on `Buffer`.
+/// * Language: Languages are crucial for selecting which OpenType feature
+///   to apply to the buffer which can result in applying language-specific
+///   behaviour. Languages are orthogonal to the scripts, and though they
+///   are related, they are different concepts and should not be confused
+///   with each other. This is controlled via the [`set_language`] method
+///   on `Buffer`.
+///
+/// Additionally, Harfbuzz can attempt to infer the values for these
+/// properties using the [`guess_segment_properties`] method on `Buffer`:
+///
+/// ```
+/// # use harfbuzz::{Buffer, Direction, sys};
+/// let mut b = Buffer::with("مساء الخير");
+/// b.guess_segment_properties();
+/// assert_eq!(b.get_direction(), Direction::RTL);
+/// assert_eq!(b.get_script(), sys::HB_SCRIPT_ARABIC);
+/// ```
+///
+/// [`set_direction`]: #method.set_direction
+/// [`set_script`]: #method.set_script
+/// [`set_language`]: #method.set_language
+/// [`guess_segment_properties`]: #method.guess_segment_properties
+pub struct Buffer {
+    /// The underlying `hb_buffer_t` from the `harfbuzz-sys` crate.
+    ///
+    /// This isn't commonly needed unless interfacing directly with
+    /// functions from the `harfbuzz-sys` crate that haven't been
+    /// safely exposed.
+    pub raw: *mut sys::hb_buffer_t,
+}
+
+impl Buffer {
+    /// Create a new, empty buffer.
+    ///
+    /// ```
+    /// # use harfbuzz::Buffer;
+    /// let b = Buffer::new();
+    /// assert!(b.is_empty());
+    /// ```
+    pub fn new() -> Self {
+        Buffer::default()
+    }
+
+    /// Create a new buffer with the given text.
+    pub fn with(text: &str) -> Self {
+        let mut b = Buffer::new();
+        b.add_str(text);
+        b
+    }
+
+    /// Create a new, empty buffer with the specified capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut b = Buffer::default();
+        b.reserve(capacity);
+        b
+    }
+
+    /// Add UTF-8 encoded text to the buffer.
+    pub fn add_str(&mut self, text: &str) {
+        unsafe {
+            sys::hb_buffer_add_utf8(
+                self.raw,
+                text.as_ptr() as *const std::os::raw::c_char,
+                text.len() as std::os::raw::c_int,
+                0,
+                text.len() as std::os::raw::c_int,
```

Comment:
The `item_offset` and `item_length` are crucial for proper use of HarfBuzz (e.g. shaping a sub-run of the text) and should be exposed in the API.

Reply:
I was thinking of exposing that via a separate function so that someone that just wants to layout a single piece of text can via something like `add_substr` instead of `add_str`. Does that make sense? IIRC, the servo code is not currently shaping substrings in a way that needs this.

Reply:
Also, thanks for the feedback! It is very helpful! I hope that the Rust binding can be easier to use than the C API with solid docs.

Reply:
I don’t speak rust, but someway to expose this should be fine I think.